readthedocs/analytics/tasks.py

```diff
@@ -80,7 +80,11 @@ def delete_old_page_counts():
     """
     retention_days = settings.RTD_ANALYTICS_DEFAULT_RETENTION_DAYS
     days_ago = timezone.now().date() - timezone.timedelta(days=retention_days)
-    return PageView.objects.filter(
-        date__lt=days_ago,
-        date__gt=days_ago - timezone.timedelta(days=90),
-    ).delete()
+    return (
+        PageView.objects.filter(
+            date__lt=days_ago,
+            date__gt=days_ago - timezone.timedelta(days=90),
+        )
+        .only("id")
+        .delete()
+    )
```

Comment:
https://docs.djangoproject.com/en/4.2/ref/models/querysets/#delete

> The delete() method does a bulk delete and does not call any delete() methods on your models. It does, however, emit the [pre_delete](https://docs.djangoproject.com/en/4.2/ref/signals/#django.db.models.signals.pre_delete) and [post_delete](https://docs.djangoproject.com/en/4.2/ref/signals/#django.db.models.signals.post_delete) signals for all deleted objects (including cascaded deletions).
>
> Django needs to fetch objects into memory to send signals and handle cascades. However, if there are no cascades and no signals, then Django may take a fast-path and delete objects without fetching into memory. For large deletes this can result in significantly reduced memory usage. The amount of executed queries can be reduced, too.
>
> ForeignKeys which are set to [on_delete](https://docs.djangoproject.com/en/4.2/ref/models/fields/#django.db.models.ForeignKey.on_delete) DO_NOTHING do not prevent taking the fast-path in deletion.

I think the real answer is to check that we don't have delete signals for any of these objects? If there are no delete signals to handle, Django will automatically follow the "fast path"?

Reply:
I did a grep in all the repositories and I didn't find any signal connected to these models.

I'm not sure this is going to solve the 30s limit issue we have when cleaning up this data, but judging from the Sentry query executed, at least, we will omit data it's not required here.

I'm not sure about what other approach we can take here; but I'm open for suggestions.

Reply:
Is it possible to do a local test and log the SQL queries from the task?

Ideally, there is just a `DELETE FROM` query. Maybe there is a `SELECT` in order to count the affected objects, but there shouldn't be a `SELECT` per row or similar expensive stuff..

Reply:
> Is it possible to do a local test and log the SQL queries from the task?

It performs a `SELECT` first. I checked this with Sentry --which is the query that takes more than 30s and logs the error. It makes sense, because it's exactly what the documentation you pasted says.

> ForeignKeys which are set to [on_delete](https://docs.djangoproject.com/en/4.2/ref/models/fields/#django.db.models.ForeignKey.on_delete) DO_NOTHING do not prevent taking the fast-path in deletion.

I'm not sure to understand this. If we add `DO_NOTHING` here will it avoid the first `SELECT`?

Reply:
I'm also confused, because I didn't find any model that has a FK to `PageView`, so there is no place to add the `DO_NOTHING` 🤷🏼

Reply:
So, the only thing that makes sense here is the `SELECT` to be able to send `pre_` and `post_` delete signals on each of these objects. That's what we need to avoid somehow.

Reply:
@humitos do you have any means of stepping through this code to see where the SELECT statements origin from? I'm thinking that _something_ prevents Django from "fast-path" deletion. Maybe not a FK relation, but maybe something else. Or some indirect FK relationship that we hadn't seen.

While this change might limit the number of columns selected, I'm guessing that the unnecessary SELECTs will remain present.